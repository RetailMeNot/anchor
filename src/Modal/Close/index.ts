export { ModalClose } from './Close.component';
