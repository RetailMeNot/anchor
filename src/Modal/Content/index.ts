export { ModalContent, StyledContent } from './Content.component';
