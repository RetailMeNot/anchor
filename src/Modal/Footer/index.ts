export { ModalFooter, StyledFooter } from './Footer.component';
