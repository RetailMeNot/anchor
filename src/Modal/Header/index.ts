export { ModalHeader, StyledHeader } from './Header.component';
