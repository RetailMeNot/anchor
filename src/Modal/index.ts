export { Modal, BaseModalBackground, ModalProvider } from './Modal.component';
