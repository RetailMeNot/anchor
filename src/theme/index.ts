export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
