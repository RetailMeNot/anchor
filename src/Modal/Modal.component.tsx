// REACT
import * as React from 'react';

// VENDOR
import * as StyledReactModal from 'styled-react-modal';
import styled, { css } from 'styled-components';
import {
    compose,
    variant,
    borderRadius as borderRadiusStyles,
    BorderRadiusProps,
    margin as marginStyles,
    MarginProps,
    color as colorStyles,
    ColorProps,
    background as backgroundStyles,
    BackgroundProps,
} from '../theme/system';
import classnames from 'classnames';

// SUBCOMPONENTS
import { ModalHeader, StyledHeader } from './Header';
import { ModalContent, StyledContent } from './Content';
import { ModalFooter, StyledFooter } from './Footer';
import { ModalClose } from './Close';

export const MODAL_THEME_KEY = 'modal';
export const MODAL_THEME = {
    size: {
        lg: {
            width: 50,
            contentPadding: 2.5,
            footerHeight: 6.5,
            height: 42.375,
        },
        sm: {
            width: 25,
            contentPadding: 2,
            footerHeight: 5.5,
            height: 42.375,
        },
    },
};

const size = variant({
    themeKey: MODAL_THEME_KEY,
    prop: 'size',
    variants: MODAL_THEME.size,
    defaultValue: 'lg',
});

const StyledModal = StyledReactModal.default.styled`
    position: relative;

    ${compose(
        marginStyles,
        backgroundStyles,
        colorStyles,
        borderRadiusStyles
    )}

    ${size(
        ({ contentPadding, footerHeight, width, height }) => css`
            height: ${({ height: propHeight }: ModalProps) =>
                propHeight || `${height}rem`};
            width: ${({ width: propWidth }: ModalProps) =>
                propWidth || `${width}rem`};

            // The Modal Content component only gets its bottom padding if
            // its the last-child of the modal. Furthermore, if the Modal Header
            // comes before it, then the Content loses its top padding
            // (see Modal Header styles). The goal is that we have full padding
            // in isolation, but lose the top or bottom padding if the Header or
            // Footer (respectively) are also used.
            ${StyledContent} {
                padding: 4rem ${contentPadding}rem 0 ${contentPadding}rem;
                &:last-child {
                    padding-bottom: ${contentPadding}rem;
                }
            }
            ${StyledHeader} {
                padding-left: ${contentPadding}rem;
            }
            ${StyledFooter} {
                padding: 0 ${contentPadding}rem;
                min-height: ${footerHeight}rem;
            }
        `
    )}

    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: space-between;
    box-shadow: ${({
        shadow = '0 0.375rem 0.5rem 0.25rem rgba(0,0,0,0.13)',
    }: ModalProps) => shadow};

    overflow: hidden;
`;

interface ModalProps
    extends StyledReactModal.ModalProps,
        MarginProps,
        BorderRadiusProps,
        ColorProps,
        BackgroundProps {
    size?: string;
    width?: string;
    height?: string;
    shadow?: string;
    children?: any;
    className?: string;
}

export const Modal = ({
    children,
    className,
    margin = 'auto',
    background = 'white',
    borderRadius = 'modal',
    ...props
}: ModalProps): React.ReactElement<ModalProps> => (
    <StyledModal
        {...{
            className: classnames('anchor-modal', className),
            margin,
            borderRadius,
            background,
            ...props,
        }}
    >
        {children}
    </StyledModal>
);

Modal.Content = ModalContent;
Modal.Header = ModalHeader;
Modal.Footer = ModalFooter;
Modal.Close = ModalClose;

interface BaseModalBackgroundProps {
    opacity?: number;
    oversized?: boolean;
    padding?: string;
}

const CustomModalBackground = styled(StyledReactModal.BaseModalBackground)`
    box-sizing: border-box;
    overflow-y: auto;
    padding: ${({ padding = '2rem 1rem' }: BaseModalBackgroundProps) =>
        padding};
    background-color: ${({ opacity = 0.6 }: BaseModalBackgroundProps) =>
        `rgba(0,0,0,${opacity})`};
`;
export const BaseModalBackground = CustomModalBackground;

export class ModalProvider extends StyledReactModal.ModalProvider {
    static defaultProps = { backgroundComponent: BaseModalBackground };
}
