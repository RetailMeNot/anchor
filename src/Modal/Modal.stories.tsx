// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import { boolean, select, number, text } from '@storybook/addon-knobs';
import styled, { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { Button, Typography } from '..';
import { RootTheme } from '../theme';
// SUBJECT
import * as README from './README.md';
import {
    Modal,
    BaseModalBackground,
    ModalProvider,
    ModalSize,
} from './Modal.component';

const { useState } = React;
const { Close, Header, Content, Footer } = Modal;

const StyledStory = styled('div')`
    background: white;
    width: 100vw;
    height: 120vh;
`;

const sizeOptions: ModalSize[] = ['sm', 'lg'];

storiesOf('Components/Modal', module)
    .addParameters({ readme: { sidebar: README } })
    .add('Default', () => {
        const OpenModalButton = () => {
            const [isOpen, setIsOpen] = useState<boolean>(true);

            return (
                <React.Fragment>
                    <Button onClick={() => setIsOpen(true)}>Open Modal</Button>
                    <Modal
                        size={select('size', sizeOptions, 'lg')}
                        isOpen={isOpen}
                        onBackgroundClick={() => setIsOpen(false)}
                        onEscapeKeydown={() => setIsOpen(false)}
                        allowScroll={boolean('allowScroll', false)}
                    >
                        <Header title="It's dangerous to go alone.">
                            <Close onClick={() => setIsOpen(false)} />
                        </Header>
                        <Content>
                            <Typography>Stuff in the middle</Typography>
                        </Content>
                        <Footer>
                            <Button
                                block
                                onClick={() => setIsOpen(false)}
                                variant="outline"
                            >
                                Decline
                            </Button>
                            <Button block onClick={() => setIsOpen(false)}>
                                Accept
                            </Button>
                        </Footer>
                    </Modal>
                </React.Fragment>
            );
        };

        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Typography tag="h5">Click this button!</Typography>
                        <OpenModalButton />
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Blank', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Modal
                            size={select('size', sizeOptions, 'sm')}
                            isOpen
                        />
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Content', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Modal size={select('size', sizeOptions, 'sm')} isOpen>
                            <Close />
                            <Content>
                                <Typography>Content Area</Typography>
                            </Content>
                        </Modal>
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Header + Content', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Modal size={select('size', sizeOptions, 'sm')} isOpen>
                            <Header title="Header Area">
                                <Close />
                            </Header>
                            <Content>
                                <Typography>Content Area</Typography>
                            </Content>
                        </Modal>
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Content + Footer', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Modal size={select('size', sizeOptions, 'sm')} isOpen>
                            <Close />
                            <Content>
                                <Typography>Content Area</Typography>
                            </Content>
                            <Footer>
                                <Typography>Footer Area</Typography>
                            </Footer>
                        </Modal>
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Oversized', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <Modal
                            isOpen
                            height={text('height', '100rem')}
                            size={select('size', sizeOptions, 'lg')}
                            backgroundProps={{
                                padding: text(
                                    'background padding',
                                    '2rem 1rem'
                                ),
                            }}
                        >
                            <Header title="Oversized Modal">
                                <Close />
                            </Header>
                            <Content>
                                <Typography>Psst... down below!</Typography>
                            </Content>
                            <Footer>
                                <Typography>Psst... up above!</Typography>
                            </Footer>
                        </Modal>
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Custom Background', () => {
        const CustomBackground = styled(BaseModalBackground)`
            background-color: rgba(255, 0, 0, 0.6);
        `;

        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider backgroundComponent={CustomBackground}>
                    <StyledStory>
                        <Modal size={select('size', sizeOptions, 'sm')} isOpen>
                            <Close />
                            <Header title="Modal with Custom Background" />
                        </Modal>
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    })
    .add('Customized', () => {
        const OpenModalButton = () => {
            const [isOpen, setIsOpen] = useState<boolean>(true);

            const pinkColorScheme = {
                base: '#DF206C',
                light: '#E665A1',
                dark: '#B81D5B',
            };

            return (
                <React.Fragment>
                    <Button onClick={() => setIsOpen(true)}>Open Modal</Button>

                    <Modal
                        isOpen={isOpen}
                        size={select('size', sizeOptions, 'sm')}
                        height={text('height', '25rem')}
                        width={text('width', '30rem')}
                        margin={text('margin', 'auto')}
                        background="linear-gradient(170deg, #DF206C 0%, #F16667 50%, white calc(50% + 1.25px))"
                        color={text('color', 'white')}
                        backgroundProps={{
                            opacity: number('backgroundProps.opacity', 0.2),
                        }}
                        allowScroll={boolean('allowScroll', false)}
                        shadow={text(
                            'shadow',
                            '0 0.375rem 0.5rem 0.25rem rgba(0,0,0,0.13)'
                        )}
                        onBackgroundClick={() => setIsOpen(false)}
                        onEscapeKeydown={() => setIsOpen(false)}
                    >
                        <Close reverse onClick={() => setIsOpen(false)} />
                        <Content>
                            <Typography
                                tag="h2"
                                scale={24}
                                weight={600}
                                align="center"
                            >
                                Customized Example
                            </Typography>
                            <Typography tag="p" align="center">
                                With some other text
                            </Typography>
                        </Content>
                        <Footer>
                            <Button
                                block
                                circular
                                variant="minimal"
                                colorTheme={pinkColorScheme}
                                onClick={() => setIsOpen(false)}
                            >
                                Decline
                            </Button>
                            <Button
                                block
                                circular
                                colorTheme={pinkColorScheme}
                                onClick={() => setIsOpen(false)}
                            >
                                Accept
                            </Button>
                        </Footer>
                    </Modal>
                </React.Fragment>
            );
        };

        return (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <StyledStory>
                        <OpenModalButton />
                    </StyledStory>
                </ModalProvider>
            </ThemeProvider>
        );
    });
