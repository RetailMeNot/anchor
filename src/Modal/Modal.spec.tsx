// REACT
import * as React from 'react';
// VENDOR
import { mount } from 'enzyme';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENT
import { Modal, ModalProvider } from './Modal.component';
import { RootTheme } from '../theme';
const { Header, Content, Footer, Close } = Modal;

describe('Component: Modal', () => {
    it('should be defined', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <div>
                        <button />
                        <Modal isOpen>Contents</Modal>
                    </div>
                </ModalProvider>
            </ThemeProvider>
        );

        const wrapper = mount(subject);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render with modal components', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ModalProvider>
                    <div>
                        <button />
                        <Modal isOpen>
                            <Header title="Header Title">
                                <Close />
                            </Header>
                            <Content>Content</Content>
                            <Footer>Footer</Footer>
                        </Modal>
                    </div>
                </ModalProvider>
            </ThemeProvider>
        );

        const wrapper = mount(subject);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
