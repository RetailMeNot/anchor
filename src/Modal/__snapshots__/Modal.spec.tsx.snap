// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Modal should be defined 1`] = `"<div><button></button></div><div><div class=\\"baseStyles__BaseModalBackground-sc-12bzkqq-0 sc-EHOje cdMXsr\\"><div class=\\"anchor-modal Modal__styles-sc-1q9xwty-0 ZRTLO\\">Contents</div></div></div>"`;

exports[`Component: Modal should render with modal components 1`] = `"<div><button></button></div><div><div class=\\"baseStyles__BaseModalBackground-sc-12bzkqq-0 sc-EHOje cdMXsr\\"><div class=\\"anchor-modal Modal__styles-sc-1q9xwty-0 ZRTLO\\"><div class=\\"anchor-modal-header sc-bwzfXH cLnCbN\\"><span class=\\"anchor-typography sc-bZQynM jpzPBl\\" scale=\\"20\\">Header Title</span><div class=\\"anchor-modal-close sc-ifAKCX eADMjw\\"><span class=\\"anchor-icon close sc-bxivhb llvVbF\\" scale=\\"lg\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\"><g fill=\\"none\\" fill-rule=\\"evenodd\\"><path d=\\"M0 0h16v16H0z\\"></path><g fill=\\"#0998D6\\" fill-rule=\\"nonzero\\"><path d=\\"M4.398 4.37l6.852 6.85-.495.495-6.852-6.85c-.292-.292-.354-.75-.073-1.032.283-.284.742-.222 1.035.07l6.85 6.851c.293.292.355.75.111.987l-.04.048c-.28.281-.74.218-1.031-.074l.495-.495c.033.034.055.044.036.078-.02.013-.027-.01-.065-.049L4.371 4.4c-.044-.043-.068-.047-.044-.072.025-.024.029 0 .071.043z\\"></path><path d=\\"M11.468 4.617l.248.248-6.851 6.85c-.292.293-.75.355-1.012.092l-.023-.021c-.281-.281-.218-.74.073-1.032l6.85-6.85c.292-.293.752-.355 1.013-.092l.023.021c.281.281.218.74-.073 1.031l-.248-.247z\\"></path></g></g></svg></span></div></div><div class=\\"anchor-modal-content sc-bdVaJa kJfdWc\\">Content</div><div class=\\"anchor-modal-footer sc-htpNat fnohta\\">Footer</div></div></div></div>"`;
